### PR TITLE
High Resolution Metrics Support

### PR DESCRIPTION
**Pull request by meshwa19**
Description of changes:

- Added support for High Resolution Metrics. Customers can specify an optional property - storageResolution of the metrics in putMetric() call. If resolution is not specified, metrics will be emitted with standard resolution.

- Added UTs and modified integration tests for the changes.

- Updated README.md and examples folder.

- Updated the canary. See the image below for memory usage. (Note: We will have to wait until the feature is enabled to get metrics in high resolution for the canary)

Screen Shot 2023-01-10 at 8 12 47 AM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -39,7 +39,7 @@ from aws_embedded_metrics import metric_scope
 @metric_scope
 def my_handler(metrics):
     metrics.put_dimensions({"Foo": "Bar"})
-    metrics.put_metric("ProcessingLatency", 100, "Milliseconds")
+    metrics.put_metric("ProcessingLatency", 100, "Milliseconds", 60)
```
Comment (rayabagi) [inline README.md:42]: Can't we pass Enum value?

File: aws_embedded_metrics/validator.py [hunk 2]
```diff
@@ -57,14 +58,15 @@ def validate_dimension_set(dimension_set: Dict[str, str]) -> None:
...
-def validate_metric(name: str, value: float, unit: Optional[str]) -> None:
+def validate_metric(name: str, value: float, unit: Optional[str], storageResolution: Optional[int], metricNameAndResolutionMap: dict) -> None:  # noqa: E501
     """
     Validates a metric
 
         Parameters:
             name (str): The name of the metric
             value (float): The value of the metric
             unit (Optional[str]): The unit of the metric
+            storageResolution (Optional[int]): The storage resolution of metric
```
Comment (rayabagi) [inline aws_embedded_metrics/validator.py:69]: Please updated the method signature parameter  for metricNameAndResolutionMap in method comments

Comment (meshwa19) [reply aws_embedded_metrics/validator.py:69]: Done!

File: aws_embedded_metrics/validator.py [hunk 3]
```diff
@@ -81,6 +84,14 @@ def validate_metric(name: str, value: float, unit: Optional[str]) -> None:
     if unit is not None and unit not in Unit:
         raise InvalidMetricError(f"Metric unit is not valid: {unit}")
 
+    if storageResolution is None or storageResolution not in StorageResolution:
+        raise InvalidMetricError(f"Metric storage Resolution is not valid: {storageResolution}")
+
+    if metricNameAndResolutionMap and name in metricNameAndResolutionMap:
+        if metricNameAndResolutionMap.get(name) is not storageResolution:
```
Comment (rayabagi) [inline aws_embedded_metrics/validator.py:91]:
```suggestion
        if if metricNameAndResolutionMap and name in metricNameAndResolutionMap and metricNameAndResolutionMap.get(name) is not storageResolution : 
```
Can't we club this into one?

File: README.md (hunk 1 above)
Comment (meshwa19) [reply README.md:42]: Done!

File: README.md [hunk 4]
```diff
@@ -67,7 +67,7 @@ Requirements:
 Examples:
 
 ```py
-put_metric("Latency", 200, "Milliseconds")
+put_metric("Latency", 200, "Milliseconds", 60)
```
[the remark this answers is not in the document]
Comment (meshwa19) [reply README.md:70]: Done!

File: tests/integ/agent/test_end_to_end.py [hunk 5]
```diff
@@ -40,7 +41,7 @@ async def test_end_to_end_tcp_multiple_flushes():
     @metric_scope
     async def do_work(metrics):
         metrics.put_dimensions({"Operation": "Agent"})
-        metrics.put_metric(metric_name, 100, "Milliseconds")
+        metrics.put_metric(metric_name, 100, "Milliseconds", StorageResolution.HIGH)
```
Comment (rayabagi) [inline tests/integ/agent/test_end_to_end.py:44]: Lets remove High. Its not required for libraries as we are not checking it against 1 sec internal in the response

Comment (meshwa19) [reply tests/integ/agent/test_end_to_end.py]: Done!

File: aws_embedded_metrics/validator.py (hunk 3 above)
Comment (meshwa19) [reply aws_embedded_metrics/validator.py]: Done

**Review by rayabagi — APPROVED** [review]
Finished with the review.

File: tests/logger/test_metrics_logger.py [hunk 6]
```diff
@@ -53,6 +54,50 @@ async def test_can_put_metric(mocker):
     assert context.metrics[expected_key].unit == "None"
 
 
+@pytest.mark.asyncio
+async def test_can_put_metric_with_different_storage_resolution_different_flush(mocker):
+    # arrange
+    expected_key = fake.word()
+    expected_value = fake.random.randrange(100)
+    metric_storageResolution = StorageResolution.HIGH
```
[the remark this answers is not in the document]
Comment (markkuhn) [reply tests/logger/test_metrics_logger.py:62]: You could also directly use `StorageResolution.HIGH` here. Don't see much purpose in assigning it to a variable first.

File: tests/serializer/test_log_serializer.py [hunk 7]
```diff
@@ -92,6 +93,52 @@ def test_serialize_metrics():
     assert_json_equality(result_json, expected)
 
 
+def test_serialize_metrics_with_Standard_storageResolution():
```
Comment (markkuhn) [inline tests/serializer/test_log_serializer.py:96]:
```suggestion
def test_serialize_metrics_with_standard_storage_resolution():
```

File: tests/serializer/test_log_serializer.py [hunk 8]
```diff
@@ -92,6 +93,52 @@ def test_serialize_metrics():
...
+    context.put_metric(expected_key, expected_value, "None", StorageResolution.STANDARD)
+
+    # act
+    result_json = serializer.serialize(context)[0]
+
+    # assert
+    assert_json_equality(result_json, expected)
+
+
+def test_serialize_metrics_with_High_storageResolution():
```
Comment (markkuhn) [inline tests/serializer/test_log_serializer.py:119]:
```suggestion
def test_serialize_metrics_with_high_storage_resolution():
```

File: README.md (hunk 1 above)
Comment (meshwa19) [reply README.md]: My bad, resolved it now. Thanks! Also added extra information regarding storage resolution.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (meshwa19) [reply README.md]: Done now.

File: aws_embedded_metrics/logger/metric.py [hunk 9]
```diff
@@ -10,12 +10,14 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+from aws_embedded_metrics.storageResolution import StorageResolution
 
 
 class Metric(object):
-    def __init__(self, value: float, unit: str = None):
+    def __init__(self, value: float, unit: str = None, storageResolution: StorageResolution = StorageResolution.STANDARD):
```
[the remark this answers is not in the document]
Comment (meshwa19) [reply aws_embedded_metrics/logger/metric.py]: Done!

File: aws_embedded_metrics/serializers/log_serializer.py [hunk 10]
```diff
@@ -87,8 +88,11 @@ def create_body() -> Dict[str, Any]:
                     if len(metric.values) > end_index:
                         remaining_data = True
 
+                metricBody = {"Name": metric_name, "Unit": metric.unit}
```
[the remark this answers is not in the document]
Comment (meshwa19) [reply aws_embedded_metrics/serializers/log_serializer.py]: Done!

File: aws_embedded_metrics/validator.py [hunk 11]
```diff
@@ -57,14 +58,16 @@ def validate_dimension_set(dimension_set: Dict[str, str]) -> None:
             raise InvalidDimensionError("Dimension name cannot start with ':'")
 
 
-def validate_metric(name: str, value: float, unit: Optional[str]) -> None:
+def validate_metric(name: str, value: float, unit: Optional[str], storageResolution: StorageResolution, metricNameAndResolutionMap: dict) -> None:  # noqa: E501
```
[the remark this answers is not in the document]
Comment (meshwa19) [reply aws_embedded_metrics/validator.py]: Done!

File: aws_embedded_metrics/validator.py [hunk 12]
```diff
@@ -81,6 +84,13 @@ def validate_metric(name: str, value: float, unit: Optional[str]) -> None:
     if unit is not None and unit not in Unit:
         raise InvalidMetricError(f"Metric unit is not valid: {unit}")
 
+    if storageResolution is None or storageResolution not in StorageResolution:
+        raise InvalidMetricError(f"Metric storage Resolution is not valid: {storageResolution}")
```
[the remark this answers is not in the document]
Comment (meshwa19) [reply aws_embedded_metrics/validator.py]: Done!

File: aws_embedded_metrics/validator.py [hunk 13]
```diff
@@ -81,6 +84,13 @@ def validate_metric(name: str, value: float, unit: Optional[str]) -> None:
...
+
+    if name in metricNameAndResolutionMap and metricNameAndResolutionMap.get(name) is not storageResolution:
+        raise InvalidMetricError(
+            "Resolution for metrics ${name} is already set. A single log event cannot have a metric with two different resolutions.")
```
[the remark this answers is not in the document]
Comment (meshwa19) [reply aws_embedded_metrics/validator.py]: Done!

File: tests/logger/test_metrics_context.py [hunk 14]
```diff
@@ -263,14 +264,16 @@ def test_put_metric_adds_metrics():
     metric_key = fake.word()
     metric_value = fake.random.random()
     metric_unit = random.choice(list(Unit)).value
+    metric_storageResolution = random.choice(list(StorageResolution)).value
```
[the remark this answers is not in the document]
Comment (meshwa19) [reply tests/logger/test_metrics_context.py]: Done!

File: tests/logger/test_metrics_logger.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (meshwa19) [reply tests/logger/test_metrics_logger.py]: Agree, removed it!

File: tests/serializer/test_log_serializer.py (hunk 7 above)
Comment (meshwa19) [reply tests/serializer/test_log_serializer.py]: Done!

File: tests/serializer/test_log_serializer.py (hunk 8 above)
Comment (meshwa19) [reply tests/serializer/test_log_serializer.py]: Done!

**Review by markkuhn — APPROVED** [review]
🚀

**Comment by hussam789** [comment]
## PR Code Suggestions ✨



<table><thead><tr><td><strong>Category</strong></td><td align=left><strong>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></td><td align=center><strong>Impact</strong></td></tr><tbody><tr><td rowspan=1>Possible issue</td>
<td>



<details><summary>Fix enum comparison logic</summary>

___


**The comparison is using identity check (<code>is not</code>) instead of equality check (<code>!=</code>) <br>for comparing enum values. This can lead to unexpected behavior when comparing <br>enum values, as they might be equal in value but not the same object instance.**

[aws_embedded_metrics/validator.py [94-96]](https://github.com/awslabs/aws-embedded-metrics-python/pull/96/files#diff-bab0a43bff403f0d35ba2a433e7b2b02cc93dd34194edd292715d66dcaf3a19dR94-R96)

```diff
-if name in metric_name_and_resolution_map and metric_name_and_resolution_map.get(name) is not storage_resolution:
+if name in metric_name_and_resolution_map and metric_name_and_resolution_map.get(name) != storage_resolution:
     raise InvalidMetricError(
         f"Resolution for metrics {name} is already set. A single log event cannot have a metric with two different resolutions.")
```

- [ ] **Apply this suggestion** 

<details><summary>Suggestion importance[1-10]: 8</summary>

__

Why: The suggestion corrects a potentially problematic identity check for enum values by using an equality check, ensuring that two equivalent enum members are recognized as equal, which improves correctness.


</details></details></td><td align=center>Medium

</td></tr>
<tr><td align="center" colspan="2">

- [ ] More 

</td><td></td></tr></tbody></table>

**Pull request merged**